src/Database/LSMTree/Internal/WriteBuffer.hs

```diff
@@ -53,55 +62,100 @@ import           Prelude hiding (lookup)
   Writebuffer type
 -------------------------------------------------------------------------------}
 
--- | The phantom type parameters provide some safety, enforcing that the types
--- of inserted entries are consistent.
---
--- TODO: Revisit this when using the write buffer from the table handle.
--- It would be consistent with other internal APIs (e.g. for @Run@ and
--- @IndexCompact@ to remove the type parameters here and move the responsibility
--- for these constraints and (de)serialisation to the layer above.
-newtype WriteBuffer =
-  WB { unWB :: Map SerialisedKey (Entry SerialisedValue SerialisedBlob) }
-  deriving stock Show
-  deriving newtype NFData
+data WriteBuffer = WB {
+    writeBufferContent   :: !(Map SerialisedKey (Entry SerialisedValue SerialisedBlob))
+    -- | Keeps track of the total size of keys and values in the buffer.
+    -- This means reconstructing the 'WB' constructor on every update.
+  , writeBufferSizeBytes :: {-# UNPACK #-} !Int
+  }
+  deriving stock (Eq, Show)
+
+instance NFData WriteBuffer where
+  rnf (WB m _) = rnf m
```

Comment:
The current instance is okay I think, but it also doesn't hurt to be extra sure (and it's slightly more future proof):
```suggestion
  rnf (WB m sb) = rnf m `seq` rnf sb
```